honeybee_radiance/primitive/geometry/cone.py

```diff
@@ -0,0 +1,180 @@
+                x1      y1      z1
+                r0      r1
+
+    """
+
+    def __init__(self, name, center_pt_start=None, radius_start=10,
+                 center_pt_end=None, radius_end=0, modifier=None, dependencies=None):
+        """Radiance Cone.
+
+        Attributes:
+            name: Geometry name as a string. Do not use white space and special
+                character.
+            center_pt_start: Cone start center point as (x, y, z) (Default: (0, 0 ,0)).
+            radius_start: Cone start radius as a number (Default: 10).
+            center_pt_end: Cone end center point as (x, y, z) (Default: (0, 0 ,10)).
+            radius_end: Cone end radius as a number (Default: 0).
+            modifier: Geometry modifier (Default: "void").
+            dependencies: A list of primitives that this primitive depends on. This
+                argument is only useful for defining advanced primitives where the
+                primitive is defined based on other primitives. (Default: [])
+        """
+        Geometry.__init__(self, name, modifier=modifier, dependencies=dependencies)
+        self.center_pt_start = center_pt_start or (0, 0, 0)
+        self.radius_start = radius_start if radius_start is not None else 10
+        self.center_pt_end = center_pt_end or (0, 0, 10)
+        self.radius_end = radius_end if radius_end is not None else 0
+        self._update_values()
+
+    def _update_values(self):
+        """update value dictionaries."""
+        self._values[2] = \
+            [self.center_pt_start[0], self.center_pt_start[1], self.center_pt_start[2],
+             self.center_pt_end[0], self.center_pt_end[1], self.center_pt_end[2],
+             self.radius_start, self.radius_end]
+
+    @property
+    def center_pt_start(self):
+        """Cone start center point as (x, y, z) (Default: (0, 0 ,0))."""
+        return self._center_pt_start
+    
+    @center_pt_start.setter
+    def center_pt_start(self, value):
+        self._center_pt_start = pv.Point3D(*(float(v) for v in value))
+
+    @property
+    def radius_start(self):
+        """Cone start radius as a number (Default: 10)."""
+        return self._radius_start
+    
+    @radius_start.setter
+    def radius_start(self, value):
+        self._radius_start = typing.float_positive(value)
+
+    @property
+    def center_pt_end(self):
+        """Cone end center point as (x, y, z) (Default: (0, 0 ,10))."""
+        return self._center_pt_end
+    
+    @center_pt_end.setter
+    def center_pt_end(self, value):
+        self._center_pt_end = pv.Point3D(*(float(v) for v in value))
+
+    @property
+    def radius_end(self):
+        """ Cone end radius as a number (Default: 0)."""
+        return self._radius_end
+    
+    @radius_end.setter
+    def radius_end(self, value):
+        self._radius_end = typing.float_positive(value)
+
+    @classmethod
+    def from_values(cls, values_dict):
+        """Make a radiance primitive from value dict.
```

Comment:
I guess the only place where it makes a difference is in the online documentation. Do you know if sphinx auto-sorts the classmethods, properties, etc?

Reply:
I don't know if there is a best practice for this. I looked it up a number of times and it looks like there is none (https://stackoverflow.com/a/36967648/4394669). I personally used to do it as \__init__, classmethods, properties, private methods and staticmethods but recently I started to push classmethods below properties as they are secondary to properties in my opinion. Now that I started to review the classes I wanted to know how this is constructed by default and what are the properties exposed and after that see what are the alternative ways to initiate it.

As you said this might be something that we can figure out in documentation. For documentation Sphix has 3 options that can be set by `autodoc_member_order`
http://www.sphinx-doc.org/en/stable/usage/extensions/autodoc.html#confval-autodoc_member_order

> This value selects if automatically documented members are sorted alphabetical (value 'alphabetical'), by member type (value 'groupwise') or by source order (value 'bysource'). The default is alphabetical.

We can start using `groupwise` and see how does that turn out to be. By default it alphabetical which is what we use currently. We can also use `bysource` to keep it in the same order as source code but then we need to standardize and enforce the order. I personally like the idea of doing it `groupwise` first and see how it will work and then go to `bysource` if we didn't like it.

Reply:
Thanks for looking into this and it is all good to know. I agree with doing the `groupwise` sorting on sphinx, though I have also been following the convention that you mentioned at the top of your response. So we are not too far away from having a standard to do it `bysource` if we wanted to.  Whatever the case, I don't know why they would ever make `alphabetical` the default.  That feels similar to David Rutten's choice of hot pink for the default preview color: Just pick something absurd so that people are forced to change it.

Reply:
> ...I have also been following the convention that you mentioned at the top of your response. So we are not too far away from having a standard to do it `bysource` if we wanted to. 

It can be the case when there are only two developers but it gets harder and harder to enforce when the project grows and honestly I don't see an obvious advantage to one over another. It is a choice of style and can be different on a case by case basis. To me it falls under (http://bikeshed.org/) and so I will use `groupwise` option so I don't have to be worried about one more thing.

Reply:
As I said before I agree, `groupwise` sounds good! I had just wanted to let you know about the option of the other one but you are right that it creates more unnecessary work to enforce it and it's not worth it to take that work on.